contracts/src/internal/HyperdrivePair.sol

```diff
@@ -0,0 +1,308 @@
+///                    particular legal or regulatory significance.
+abstract contract HyperdrivePair is
+    IHyperdriveEvents,
+    HyperdriveBase,
+    HyperdriveMultiToken
+{
+    using FixedPointMath for uint256;
+    using FixedPointMath for int256;
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    // FIXME: Do we need a `minOutput` parameter here? It feels kind of paranoid,
+    // but it's probably a good idea with things like prepaid interest.
+    //
+    /// @dev Mints a pair of long and short positions that directly match each
+    ///      other. The amount of long and short positions that are created is
+    ///      equal to the base value of the deposit. These positions are sent to
+    ///      the provided destinations.
+    /// @param _amount The amount of capital provided to open the long. The
+    ///        units of this quantity are either base or vault shares, depending
+    ///        on the value of `_options.asBase`.
+    /// @param _options The pair options that configure how the trade is settled.
+    /// @return maturityTime The maturity time of the new long and short positions.
+    /// @return bondAmount The bond amount of the new long and short positoins.
+    function _mint(
+        uint256 _amount,
+        uint256 _minVaultSharePrice,
+        IHyperdrive.PairOptions calldata _options
+    )
+        internal
+        nonReentrant
+        isNotPaused
+        returns (uint256 maturityTime, uint256 bondAmount)
+    {
+        // Check that the message value is valid.
+        _checkMessageValue();
+
+        // Check that the provided options are valid.
+        _checkPairOptions(_options);
+
+        // Deposit the user's input amount.
+        (uint256 sharesDeposited, uint256 vaultSharePrice) = _deposit(
+            _amount,
+            _options.asBase,
+            _options.extraData
+        );
+
+        // Enforce the minimum transaction amount.
+        //
+        // NOTE: We use the value that is returned from the deposit to check
+        // against the minimum transaction amount because in the event of
+        // slippage on the deposit, we want the inputs to the state updates to
+        // respect the minimum transaction amount requirements.
+        //
+        // NOTE: Round down to underestimate the base deposit. This makes the
+        //       minimum transaction amount check more conservative.
+        if (
+            sharesDeposited.mulDown(vaultSharePrice) < _minimumTransactionAmount
+        ) {
+            revert IHyperdrive.MinimumTransactionAmount();
+        }
+
+        // Enforce the minimum vault share price.
+        if (vaultSharePrice < _minVaultSharePrice) {
+            revert IHyperdrive.MinimumSharePrice();
+        }
+
+        // Perform a checkpoint.
+        uint256 latestCheckpoint = _latestCheckpoint();
+        uint256 openVaultSharePrice = _applyCheckpoint(
+            latestCheckpoint,
```

Comment:
@jalextowle Quick question: why here the `latestCheckpoint` gets passed in as the 1st input param? The `_applyCheckpoint(uint256 _checkpointTime,uint256 _vaultSharePrice,uint256 _maxIterations,bool _isTrader)` says "@param _checkpointTime The time of the checkpoint to create.", so intuitively shouldn't it be `block.timestamp` or something?